oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/OAuth2TokenAccessor.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.core;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.springframework.util.Assert;
+
+/**
+ * An OAuth 2.0 token that implements {@link ClaimAccessor}
+ *
+ * @author Josh Cummings
+ * @since 5.2
+ */
+public class OAuth2TokenAccessor extends AbstractOAuth2Token implements ClaimAccessor {
```

Comment:
Some ideas for a name `OAuth2TokenAttributes` `OAuth2TokenClaims`

Reply:
I like `OAuth2TokenClaims` which may also be the parent for `Jwt` and we can also considerate it for `OidcIdToken`. However, what do we name the `Authentication` class? If we follow the same naming pattern for other `AbstractAuthenticationToken` implementations than we have `OAuth2TokenClaimsAuthenticationToken`. Same would apply to `OAuth2TokenAttributesAuthenticationToken`.

@rwinch This brings us back to the same naming issue  you pointed out [here](https://github.com/spring-projects/spring-security/pull/6352#discussion_r250299115)

Reply:
@jzheaux One thing I want to point out when you do the comparison between `UserDetails` and the potential/proposed name `OAuth2TokenDetails`

> A UserDetails contains the user and details about the user. "OAuth2TokenClaims contains the token and claims about the token" doesn't feel as clear.

It's not always the case that a user is associated to an access token. For the `client_credentials` grant, there never is a user association. So `OAuth2TokenDetails` would not make sense here as a comparison to `UserDetails`.

Reply:
My comparison is not based on the merit of it being a "User". Indeed, there isn't really a user in many requests with Resource Server. It is based on the semantics of the word "Details".

Because `UserDetails` implies that the details can include the `User`, too, then `OAuth2TokenDetails` implies that the details can include the `OAuth2Token`. `Claims` doesn't bear this same weight in the codebase. I don't believe users would infer that `OAuth2TokenClaims` would contain anything more than the claims themselves.

Reply:
Another idea is `DetailedOAuth2Token`. It indicates that this is a token with additional information. `StructuredOAuth2Token` contains spec language, e.g. the introspection endpoint returns the token in a JSON structure, the JWT spec describes the token in the JWT structure, and I like it for that reason.

Reply:
This is resolved via https://github.com/spring-projects/spring-security/issues/6498